src/lib/components/Flowbite/FlowbiteThemeProvider.tsx

```diff
@@ -0,0 +1,12 @@
+import { FC, ReactNode } from 'react';
+import { DeepPartial } from '..';
+import { Flowbite, FlowbiteTheme } from './Flowbite';
+
+interface ThemeProviderProps {
+  children: ReactNode;
+  theme: DeepPartial<FlowbiteTheme>;
+}
+
+export const FlowbiteThemeProvider: FC<ThemeProviderProps> = ({ children, theme }) => (
+  <Flowbite theme={{ theme }}>{children}</Flowbite>
```

Comment:
I'm not sure about this... The `Flowbite` component handles not just the component styling, but also, dark and light mode. I'm afraid that it will cause a lot of "hooks" and extra processing under the hood.

Reply:
What do you think @tulup-conner?

Reply:
I think that make sense, also we can create object and then pass it to props and this way seems clear
```JSX
const theme = {
  avatar: {
    size: {
      md: 'w-6 h-6',
    },
  },
};

const Test = () => (
  <Flowbite theme={{ theme }}>
    <div />
  </Flowbite>
);
```

Reply:
I'm not convinced that using `Flowbite` everywhere or `FlowbiteThemeProvider` is the best approach. But I could be totally wrong. The initial idea to pass to the components the `theme=` was from @tulup-conner, so... I really would like to get his opinion about it.

What are the pros and cons of both approaches? How many renderings or hooks do we get? How does it impact the performance? AFAIK, provider context could be quite expensive: [check here](https://javascript.plainenglish.io/should-a-context-provider-be-that-expensive-7cafa3727507) and also [here](https://blog.logrocket.com/pitfalls-of-overusing-react-context/).

But I'm not a front-end expert. 😶‍🌫️ 

From a code perspective, I feel quite confused about wrapping every component with the context provider in order to pass the theme to it. OK, it could look nice if you have multiple components that you need to customize in a section, but... Still, feels weird to have an "application theme context" and then inside this context have another "theme context" to overwrite one, two, or more components themes.

But this is just my opinion, I would like to hear some pros and cons of both approaches, so we could take a solid decision here. I felt like this is a really important and nice feature either way, and it will help the flowbite-react community a lot.

Reply:
- I've only thought having theme-per-component because I know it would be useful; I haven't reasoned through it at all until now. 

With that said, this

```js
<Flowbite theme={..}>
  <Component props />
</Flowbite>
```

Is more verbose than

```js
<Component props theme={..} />
```

Critically, the theme context solution also adds a level of nesting, which is just not ideal ergonomically IMO.

With theme-per-component, in situations where you would like to reuse a `flowbite-react` component with a specific custom theme, you just extract that line to its own component. 

```js
const ThemedNavbar: FC = function () {
  return (
    <Navbar props theme={..} />
  )
}
```

This is the same methodology encouraged by the underlying technology, Tailwind CSS, in situations where you find yourself writing long-winded `className`s.

Like @rluders said, the existing design of this `Flowbite` component serves a few different purposes, including the global dark/light theme setting. We would need to change this implementation further to extract the theme to its own component and context, presumably. I can't imagine a situation where having multiple contexts with their own idea of the current theme setting is useful.

- I think the decision to couple the `FlowbiteTheme` pieces to their actual components is beyond sensible. We should definitely do that, IMO.
- I agree we should update `mergeDeep` since JS is [call-by-sharing](https://pmg.csail.mit.edu/papers/thetaref/node34.html). I misunderstood the behavior. Thanks, I just learned about this because of your PR!
- On performance, I think people generally have a habit of prematurely optimizing things. We would want to look for metrics to determine whether performance *meaningfully* degrades with the `Flowbite` context providers, and how many providers need to exist before degradation starts. Tangentially, the inline theme solution also has the benefit of not being React Context - just a prop.

Reply:
> - I think the decision to couple the FlowbiteTheme pieces to their actual components is beyond sensible. We should definitely do that, IMO.

+1 to this... I think that we should definitely do it, no matter if we decide to go for per-component theming or not, this is definitely a huge benefit and introduces a lot to maintainability and readability to the codebase.

> - On performance, I think people generally have a habit of prematurely optimizing things. We would want to look for metrics to determine whether performance meaningfully degrades with the Flowbite context providers, and how many providers need to exist before degradation starts. Tangentially, the inline theme solution also has the benefit of not being React Context - just a prop.

I don't want to look for premature optimization here, I think that the point was that given its complexity, and from what I read using Context could be a little bit trickier to use.

I would definitely aim for simplicity, and use the `theme` property per component. It would give us the flexibility that we are looking for to introduce an easier and more solid way to customize flowbite-react components without introducing unnecessary verbosity and under-the-hood processes.

So, @sldk-yuri how about we first aim to make all components have their own theme interface?

Reply:
Thank you for you opinions @rluders @tulup-conner. Nested context is a one of the ways that we can use, probably the simpliest in current situation, becaue we don't have to refactor/update all components with theme prop to aplly theme ovveriding, just use wrapper. But in the same case deep context nesting could be a potential problem in the future and probably it's over-solution right now. Also we may add this feature in the future if needed. I'm going to start with `theme` prop option.

About this PR, I'm going to revert branch to f45cecf393f8f2407faa1134d2a9b3b260eb4fb8 because this commit remove nested providers feature c4e195c831712e85c8be2edce8a02f13ce52569b and save:

- interface decomposition 1acaae84fe83e5a5e616380209f8d66ba440e067, 
- moves interfaces to their components 1ac4e68f1b2ad02b51012b77e46ff4990f37248f, 
- test cases for Flowbite provider 95e02fbcbd5b031f84fdb69bc72824e7b0c84853 
- deep Merge fix f45cecf393f8f2407faa1134d2a9b3b260eb4fb8.

Should we merge this PR with these 4 commits and open new one for components `theme` prop or should I continue to work here?